### PR TITLE
Listen to 127.0.0.1 if node ip is 127.0.0.1

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
This should fixed the new mac popups in #18730

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]

**Comment by jjyao** [comment]
The failed tests are not related. Ready to be merged.

File: dashboard/agent.py [hunk 1]
```diff
@@ -83,9 +83,10 @@ def __init__(self,
             assert self.ppid > 0
             logger.info("Parent pid is %s", self.ppid)
         self.server = aiogrpc.server(options=(("grpc.so_reuseport", 0), ))
+        grpc_ip = "127.0.0.1" if self.ip == "127.0.0.1" else "0.0.0.0"
```
Comment (rkooo567) [inline dashboard/agent.py:86]: I wonder if we should just do `platform.system() == "Darwin"` instead?

**Review by rkooo567 — APPROVED** [review]
This looks good. Btw, is there any good regression test we can verify there's no MacOS popup? Like having some sanity check on port numbers to make sure it is not 0.0.0.0 for example?

File: dashboard/agent.py (hunk 1 above)
Comment (jjyao) [reply dashboard/agent.py:86]: If node ip is 127.0.0.1, then we are in the local/single node mode regardless if it's linux, mac or windows, and we should bind to 127.0.0.1

And we have a central place to decide the node ip which is `services.get_node_ip_address` so if we want to do the same thing for other platform, we just need to change that one place. In fact, both windows and mac have the popup issue.

Comment (fyrestone) [reply dashboard/agent.py:86]: We should keep the dashboard agent listen ip the same as raylet, if the raylet listens to 127.0.0.1, then the dashboard agent needs to listen to 127.0.0.1

File: dashboard/head.py [hunk 2]
```diff
@@ -117,9 +117,10 @@ def __init__(self, http_host, http_port, http_port_retries, redis_address,
         ip, port = redis_address.split(":")
         self.gcs_client = connect_to_gcs(ip, int(port), redis_password)
         self.server = aiogrpc.server(options=(("grpc.so_reuseport", 0), ))
+        grpc_ip = "127.0.0.1" if self.ip == "127.0.0.1" else "0.0.0.0"
         self.grpc_port = ray._private.tls_utils.add_port_to_grpc_server(
-            self.server, "[::]:0")
-        logger.info("Dashboard head grpc address: %s:%s", self.ip,
+            self.server, f"{grpc_ip}:0")
+        logger.info("Dashboard head grpc address: %s:%s", grpc_ip,
```
Comment (fyrestone) [inline dashboard/head.py:123]: Is the bind ip the same as `--dashboard-host` in https://github.com/ray-project/ray/blob/master/python/ray/scripts/scripts.py#L374?

File: dashboard/agent.py (hunk 1 above)
Comment (jjyao) [reply dashboard/agent.py:86]: Yes, ralyet has the same logic so they are consistent.

File: dashboard/head.py (hunk 2 above)
Comment (jjyao) [reply dashboard/head.py:123]: seems may not be the case: `self.ip = ray.util.get_node_ip_address()`

File: python/ray/tests/test_basic_3.py [hunk 3]
```diff
@@ -315,5 +318,35 @@ def g():
     assert ray.worker.global_worker.redis_client.llen("Exports") == num_exports
 
 
+@pytest.mark.skipif(
+    sys.platform not in ["win32", "darwin"],
+    reason="Only listen on localhost by default on mac and windows.")
+def test_listen_on_localhost(ray_start_regular):
```
Comment (rkooo567) [inline python/ray/tests/test_basic_3.py:324]: Nice test!

**Review by rkooo567 — APPROVED** [review]
@fyrestone can I merge this PR?

**Pull request merged**